src/react-router/useSearchParams.ts

```diff
@@ -0,0 +1,77 @@
+import { useCallback, useEffect, useMemo, useRef } from "react"
+import { useLocation, useNavigate } from "react-router"
+
+type SearchParamsSetterValue =
+  | URLSearchParams
+  | ((prevSearchParams: URLSearchParams) => URLSearchParams)
+
+type SetSearchParmas = (newSearchParams: SearchParamsSetterValue) => void
+
+/**
+ * A hook for getting and setting URL Search Parameters with React Router.
+ *
+ * NOTE: React Router v6 has a built-in hook for this, but it has an issue with
+ * multiple updates.
+ *
+ * With React's own `useState` hook, updater functions are passed the CURRENT
+ * value. For example:
+ * ```ts
+ * const [count, setCount] = useState(0)
+ * const incrementBy2 = () => {
+ *    setCount((prevCount) => prevCount + 1)
+ *    setCount((prevCount) => prevCount + 1)
+ * }
+ * ```
+ * Calling `incrementBy2` will increment `count` by 2: the second call to
+ * `setCount` will use the updated value from the first call.
+ *
+ * React Router's own `useSearchParams` hook does not have this behavior: it
+ * uses outdated values when multiple updates occur in a row. This inhibits
+ * the ability independently update multiple search params in a single render
+ * cycle.
+ */
+const useSearchParams = (): [URLSearchParams, SetSearchParmas] => {
```

Comment:
This works nicely, but can it be done internally to `@mitodl/course-search-utils`?, ie. `hooks/useSearchQueryParams.ts` imports from `react-router/useSearchParams.ts` and the passed `setSearchParams` calls into it. This is to avoid importing `useSearchParams` only to pass their returns back into `useSearchQueryParams()`. I have reservations about adding `react-router` as a dependency rather than just passing the initial `location` (or search state) and some `onSearchChanged` callback.

Reply:
My main rationale for making `useSearchQueryParams` take ` { searchParams, setSearchParams }` as args rather than calling `useSearchParams` internally was to avoid making `react-router` a requirement. In this way, the course-search-utils hook could be used with either [NextJS's useSearchParams](https://nextjs.org/docs/app/api-reference/functions/use-search-params) or [React Router's useSearchParams](https://reactrouter.com/en/main/hooks/use-search-params) (except if you do use react-router's, you need to deal with the bug mentioned here), or you could use it with `useState(new URLSearchParams())`.

All that said, React Router **is** what our projects currently use, so calling it internally would be reasonable.

Reply:
Actually, with NextJS I think you'd need to use a combination of their [`useSearchParams`](https://nextjs.org/docs/app/api-reference/functions/use-search-params) (which is read-only) and [`useRouter`](https://nextjs.org/docs/app/api-reference/functions/use-router) (for updating).